indica/readers/adas.py

```diff
@@ -183,7 +186,175 @@ def get_adf15(
             different charge state.
 
         """
-        return DataArray()
+
+        def build_file_component(year, element):
+            file_component = "transport"
+            if year != "transport":
+                file_component = f"pec{year}][{element.lower()}"
+
+            return file_component
+
+        def file_type(identifier):
+            identifier_dict = {
+                "+": "compact",
+                ":": "expanded",
+            }
+            file_type = identifier_dict.get(identifier)
+            if file_type is None:
+                raise ValueError(f"Unknown file header identified ({identifier}).")
+
+            return file_type
+
+        def transition_match(transition_line):
+            transition_type = "orbitals"
+            match = (
+                r"c\s+(\d+.)"  # isel
+                r"\s+(\d+.\d+)"  # wavelength
+                r"\s+(\d+)(\(\d\)\d\(.+\d?.\d\))-"  # transition upper level
+                r".+(\d+)(\(\d\)\d\(.+\d?.\d\))"  # transition lower level
+            )
+            header_re = re.compile(match)
+            m = header_re.search(transition_line)
+            if not m:
+                transition_type = "n_levels"
+                match = r"c\s+(\d+.)\s+(\d+.\d+)\s+([n]\=.\d+.-.[n]\=.\d+)"
+                header_re = re.compile(match)
+                m = header_re.search(transition_line)
+                if not m:
+                    raise ValueError(f"Unknown transition formatting ({identifier}).")
```

Comment:
Perhaps worth including the line that failed to match in the exception message to make errors quicker to debug? Up to you if you want to keep `identifier` as well or replace it

Reply:
Seems like this requires the **readlines** method, so either we change to readlines or I'm not sure how this can be done.
It's anyway not too complicated to debug this code, so I have no problems in leaving it as is.

Reply:
The `transition_line` argument is the line currently being matched so could just include that. Not desparate though so I won't hold up the merge because of it